python/cuml/test/test_nearest_neighbors.py

```diff
@@ -315,3 +318,40 @@ def test_knn_graph(input_type, nrows, n_feats, p, k, metric, mode,
         assert cupyx.scipy.sparse.isspmatrix_csr(sparse_cu)
     else:
         assert isspmatrix_csr(sparse_cu)
+
+
+@pytest.mark.parametrize("metric", valid_metrics(cuml_algo="sparse"))
+@pytest.mark.parametrize('nrows', [10, 35])
+@pytest.mark.parametrize('ncols', [10, 35])
+@pytest.mark.parametrize('density', [0.8, 0.9])
+@pytest.mark.parametrize('n_neighbors', [2, 4])
+@pytest.mark.parametrize('batch_size_index', [10, 20000])
```

Comment:
This is a pretty big matrix of tests, some of which are fairly big. It's worth thinking through whether we can both shrink it and push more on corner cases. Couple of thoughts:
 * Maybe an explicit parametrized list rather than a full cartesian product?
 *  I don't see _output_cumlarray tested
 * Some corner cases such as nrows=1, ncols=1, ncols > nrows, n_neighbors=1.
 * Not sure the two different densities make a big enough difference to do a full matrix of them?

Reply:
> This is a pretty big matrix of tests, some of which are fairly big. 

The batch_size of `20000` in the case of this test just turns off the batching. The data sizes are very small (nnz of 35 * 35 * 0.9 at the largest). 

> I don't see _output_cumlarray tested

This gets call internally by `test_knn_graph`. I changed the `_output_cumlarray` to `_output_type`, which makes the logic correct. 

 > Not sure the two different densities make a big enough difference to do a full matrix of them?

That is true- I was having problems getting the tests to pass w/ smaller densities. The increased sparsity provided more opportunity for strange determinism errors in the order of the resulting indices. Maybe a reasonable trade-off for now is just to use 0.8.  

> Some corner cases such as nrows=1, ncols=1, ncols > nrows, n_neighbors=1.

I implemented the `nrows=1`, `n_neighbors=1`, and `ncols > nrows` but `ncols=1` seems to have some trouble with the sparsity and non-determinism when computing the distances.